### DISCUSSION TITLE
What are the necessary elements for creating a new node

### DISCUSSION THREAD

**Discussion by PlethoraChutney** [BradyAJohnston/MolecularNodes#262] [General]
Let's say, hypothetically, that someone wanted to write a new node that let you select atoms/residues/what have you by their ChimeraX selection string. That person might be confused with where to start writing a node.

It seems that the business logic for a node involves building a node tree in `nodes.py`. For instance, the chain selection logic lives in `nodes.py`, lines 446--534.

To get this new node to appear, you then need to add two things to `ui.py`. First, a `layout_function` subclass which describes how the menu looks, and a `bpy.types.Operator` subclass which describes what happens. This would be `menu_chain_selection_custom` and `MOL_OT_Chain_Selection_Custom`, respectively.

This is where it gets confusing for me. Modifying what you've got in (e.g.) `menu_chain_selection_custom` does modify what is displayed in the addon, but trying to add a new box or whatever doesn't work. Working on PDB caching made me wonder if I *also* would need to add a `bpy.props.StringProperty` to `load.py`?

But then I looked for `MOL_color_element` and I can't even find where that's defined at all!

I guess my question is: *for any given node, what properties need to be added to what files for what functionality*. I've tried googling around but it seems like a lot of blender addons are just single-file `__init__.py` things, which muddy the waters.

Thank you in advance!

**Comment by BradyAJohnston** [comment]
The majority of the nodes don't involve python at all, but because GN is still in its infancy, things that involved loops and text parsing all require some calls to python scripts.

## Creating New Nodes

There are two types of nodes that I have created so far.

The majority of the nodes are pre-made by hand, and live inside of the `assets/node_append_file.blend` file that is in the repo. This gets shipped with the addon and when the user clicks the button the pre-made node is just appended to the file and then added to the node tree. 

The other types of nodes are the ones that need to be structure-specific, so basically anything that would require a loop. You can't loop currently inside of GN, so in creating those nodes I have to make the loops myself when generating them using python. Those are the majority of the nodes defined inside of `nodes.py`. This enables the chain selection nodes that are the correct size for the number of chains for the structure, and to have the names correct etc.

`MOL_color_element` is one of the pre-defined nodes (for the first 20 elements, just for convenience) so it doesn't exist inside of the `nodes.py`. Selections for chains & ligands, as well as colouring by chain, are all created procedurally based on the information from the structure for the number and names of the chains.

## A new custom selection string node

There isn't yet useful text parsing inside of GN, so the text needs to be parsed by python then a custom node group created. If you look at `resid_multiple_selection()` function in `nodes.py` this does exactly this. You can input a custom `resid` selection string, and it will build a custom node tree that creates the selection. It can be used by clicking `MolecularNodes -> Selections -> Res ID` and a custom selection input pops up, which allows for the custom numerical ranges and the resulting nodes that are created:

![image](https://github.com/BradyAJohnston/MolecularNodes/assets/36021261/3b3f8d2a-1484-4fdf-90ca-3659f97fd0b9)

![image](https://github.com/BradyAJohnston/MolecularNodes/assets/36021261/112ecc40-3edf-4234-935d-cd5880a35e38)

![image](https://github.com/BradyAJohnston/MolecularNodes/assets/36021261/d72fd46e-b09f-4d44-9843-bebd82c50eea)

If you want to implement a new custom selection that would mimic that of ChimeraX, I would certain welcome it but it might be a bit finnicky. You would have to write your own text parses to take the inputted text, then createa  new node group which combines the boolean selections of `chain_id`, `res_id` and `res_name` etc. You could copy the operator `MOL_OT_Residues_Selection_Custom` which provides the popup then executes the function for creating the node.

It's kind of all over the place, and ideally at some point text parsing will be possible inside of GN which will avoid this back and forth, but currently I believe this is what we are limited to doing.

Hopefully this clears up some of how the logic of the nodes work. Happy to talk about it plenty more!

**Reply by PlethoraChutney** [reply]
Ah, awesome. Thanks so much!
